c2rust-analyze/src/dataflow/type_check.rs

```diff
@@ -299,24 +299,28 @@ impl<'tcx> TypeChecker<'tcx, '_> {
 
                     Some(Callee::Trivial) => {}
 
-                    Some(Callee::Malloc) => {
-                        let destination = self.acx.c_void_casts.get_or_default_to(&destination);
-                        self.visit_place(destination, Mutability::Mut);
-                    }
-
-                    Some(Callee::Calloc) => {
-                        let destination = self.acx.c_void_casts.get_or_default_to(&destination);
-                        self.visit_place(destination, Mutability::Mut);
+                    Some(Callee::Malloc | Callee::Calloc) => {
+                        let out_ptr = self
+                            .acx
+                            .c_void_casts
+                            .get_adjusted_place(CVoidCastDirection::From, destination);
+                        self.visit_place(out_ptr, Mutability::Mut);
                     }
                     Some(Callee::Realloc) => {
-                        let destination = self.acx.c_void_casts.get_or_default_to(&destination);
-                        let input_ptr_pl = args[0].place().unwrap();
-                        let first_arg = self.acx.c_void_casts.get_or_default_to(&input_ptr_pl);
-                        self.visit_place(destination, Mutability::Mut);
-                        let pl_lty = self.acx.type_of(destination);
+                        let out_ptr = self
+                            .acx
+                            .c_void_casts
+                            .get_adjusted_place(CVoidCastDirection::From, destination);
+                        let in_ptr = args[0].place().unwrap();
+                        let in_ptr = self
+                            .acx
+                            .c_void_casts
+                            .get_adjusted_place(CVoidCastDirection::To, in_ptr);
+                        self.visit_place(out_ptr, Mutability::Mut);
+                        let pl_lty = self.acx.type_of(out_ptr);
                         assert!(args.len() == 2);
-                        self.visit_place(input_ptr_pl, Mutability::Not);
-                        let rv_lty = self.acx.type_of(first_arg);
+                        self.visit_place(in_ptr, Mutability::Not);
```

Comment:
i think the semantics of this have changed. in the previous version, the original argument to realloc was visited, and here it seems to be the adjustment that's visited. is this intentional? if not, check to see if a similar change exists for the other callees. something about this feels more correct -- we're avoiding visiting a place that will cease to be relevant given that we're skipping the cast. but i wonder if we risk traversing the adjustment, the new place, superfluously and if that has any unintended side effects

Reply:
You're right, I accidentally changed it.  But I was under the impression they should be the same (i.e. I meant to do this, but the change was accidental; I didn't realize it was different before).

See https://github.com/immunant/c2rust/pull/826/files#r1106252674 also.

Reply:
let's revive the old semantics until we determine that a change to them is desired/necessary